src/test/java/org/paukov/combinatorics3/KPermutationTest.java

```diff
@@ -0,0 +1,302 @@
+
+  @Test
+  public void test_simple_with_equal_elements_k3_permutation() {
+    List<List<Integer>> permutations =
+            Generator.permutation(Arrays.asList(1, 2, 2, 3))
+                    .k(3)
+                    .stream()
+                    .collect(toList());
+
+    assertThat(permutations).hasSize(12);
+
+    System.out.println("Simple with equal elements k3-permutations of (1, 2, 2, 3):");
+    permutations.stream().forEach(System.out::println);
+
+    assertThat(permutations.get(0)).containsExactly(1, 2, 2);
+    assertThat(permutations.get(1)).containsExactly(2, 1, 2);
+    assertThat(permutations.get(2)).containsExactly(2, 2, 1);
+    assertThat(permutations.get(3)).containsExactly(1, 2, 3);
+    assertThat(permutations.get(4)).containsExactly(1, 3, 2);
+    assertThat(permutations.get(5)).containsExactly(3, 1, 2);
+    assertThat(permutations.get(6)).containsExactly(3, 2, 1);
+    assertThat(permutations.get(7)).containsExactly(2, 3, 1);
+    assertThat(permutations.get(8)).containsExactly(2, 1, 3);
+    assertThat(permutations.get(9)).containsExactly(2, 2, 3);
+    assertThat(permutations.get(10)).containsExactly(2, 3, 2);
+    assertThat(permutations.get(11)).containsExactly(3, 2, 2);
+  }
+
+  @Test
+  public void test_simple_permutation_iterator_remove_operation() {
+    Iterator<List<Integer>> permutations =
+        Generator.permutation(Arrays.asList(1, 2, 3))
+            .k(2)
+            .iterator();
+
+    assertThat(permutations).isNotNull();
+    assertThat(permutations.hasNext()).isTrue();
+
+    // this method should throw a UnsupportedOperationException
+    assertThrows(UnsupportedOperationException.class, permutations::remove);
+  }
+
+  @Test
+  public void test_simple_k_permutations_iterator_toString() {
+    Iterator<List<Integer>> permutations =
+        Generator.permutation(Arrays.asList(1, 2, 3))
+            .k(2)
+            .iterator();
+
+    assertThat(permutations).isNotNull();
+    assertThat(permutations.hasNext()).isTrue();
+    assertThat(permutations.next()).containsExactly(1, 2);
+  }
+
+  @Test
+  public void test_duplicated_k_permutation_iterator_remove_operation() {
+    Iterator<List<Integer>> permutations =
+        Generator.permutation(Arrays.asList(1, 2, 2))
+            .k(2)
+            .iterator();
+
+    assertThat(permutations).isNotNull();
+    assertThat(permutations.hasNext()).isTrue();
+
+    // this method should throw a UnsupportedOperationException
+    assertThrows(UnsupportedOperationException.class, permutations::remove);
+  }
+
+  @Test
+  public void test_duplicated_k_permutations_iterator_toString() {
+    Iterator<List<Integer>> permutations =
+        Generator.permutation(Arrays.asList(1, 2, 2))
+            .k(2)
+            .iterator();
+
+    assertThat(permutations).isNotNull();
+    assertThat(permutations.hasNext()).isTrue();
+    assertThat(permutations.next()).containsExactly(1, 2);
+    assertThat(permutations.next()).containsExactly(2, 1);
+  }
```

Comment:
test_duplicated_k_permutation_iterator_remove_operation()
test_duplicated_k_permutations_iterator_toString()

These tests should be also removed, they do not test the iterator methods here.

Reply:
Done